pkg/scheduler/plugins/elasticquota/plugin.go

```diff
@@ -0,0 +1,560 @@
+}
+
+func (g *Plugin) Reserve(ctx context.Context, state *framework.CycleState, p *corev1.Pod, nodeName string) *framework.Status {
+	g.Lock()
+	defer g.Unlock()
+
+	quotaName := g.getPodAssociateQuotaName(p)
+	// update groupQuotaManager
+	quotaInfo := g.groupQuotaManager.GetQuotaInfoByName(quotaName)
+	podRequest, _ := resource.PodRequestsAndLimits(p)
+	g.groupQuotaManager.UpdateGroupDeltaUsed(quotaInfo.Name, podRequest)
+	// update elasticQuotaInfos
+	elasticQuotaInfo := g.elasticQuotaInfos[quotaName]
+	if elasticQuotaInfo != nil {
+		err := elasticQuotaInfo.addPodIfNotPresent(p)
+		if err != nil {
+			klog.ErrorS(err, "Failed to add Pod to its associated elasticQuota", "pod", klog.KObj(p))
+			return framework.NewStatus(framework.Error, err.Error())
+		}
+	}
+	// update quotaOverUsedRevokeController
+	g.quotaOverUsedRevokeController.addPodWithQuotaName(quotaName, p)
+	return framework.NewStatus(framework.Success, "")
+}
+
+func (g *Plugin) Unreserve(ctx context.Context, state *framework.CycleState, p *corev1.Pod, nodeName string) {
+	g.Lock()
+	defer g.Unlock()
+
+	quotaName := g.getPodAssociateQuotaName(p)
+	// update groupQuotaManager
+	quotaInfo := g.groupQuotaManager.GetQuotaInfoByName(quotaName)
+	podRequest, _ := resource.PodRequestsAndLimits(p)
+	podRequestTmp := quotav1.Subtract(corev1.ResourceList{}, podRequest)
+	g.groupQuotaManager.UpdateGroupDeltaUsed(quotaInfo.Name, podRequestTmp)
+	// update elasticQuotaInfo
+	elasticQuotaInfo := g.elasticQuotaInfos[quotaName]
+	if elasticQuotaInfo != nil {
+		err := elasticQuotaInfo.deletePodIfPresent(p)
+		if err != nil {
+			klog.ErrorS(err, "failed to add Pod to its associated elasticQuota", "pod", klog.KObj(p))
+		}
+	}
+	g.quotaOverUsedRevokeController.erasePodWithQuotaName(quotaName, p)
+}
+
+func (g *Plugin) RevokePodDueToQuotaOverUsed() {
+	toRevokePods := g.quotaOverUsedRevokeController.MonitorAll()
+	for _, pod := range toRevokePods {
+		quotaName := g.getPodAssociateQuotaName(pod)
+		if err := util.DeletePod(g.handle.ClientSet(), pod); err != nil {
+			klog.Errorf("failed to revoke pod due to quota overused, pod:%v, quota:%v, error:%s",
+				pod.Name, quotaName, err)
+			continue
+		}
+		klog.V(5).Infof("finish revoke pod due to quota overused, pod:%v, quota:%v",
+			pod.Name, quotaName)
+	}
+}
+
+func (g *Plugin) OnQuotaAdd(obj interface{}) {
+	quota, ok := obj.(*schedulerv1alpha1.ElasticQuota)
+	if !ok {
+		klog.Errorf("quota is nil")
+		return
+	}
+
+	if quota.DeletionTimestamp != nil {
+		klog.Errorf("quota is deleting:%v.%v", quota.Namespace, quota.Name)
+		return
+	}
+
+	g.Lock()
+	defer g.Unlock()
+
+	if oldQuotaInfo := g.groupQuotaManager.GetQuotaInfoByName(quota.Name); oldQuotaInfo != nil && quota.Name != extension.DefaultQuotaName {
+		return
+	}
+
+	klog.V(5).Infof("OnQuotaAddFunc add quota:%v.%v", quota.Namespace, quota.Name)
+
+	g.elasticQuotaInfos[quota.Name] = NewSimpleQuotaInfo(quota.Name)
+	err := g.groupQuotaManager.UpdateQuota(quota, false)
```

Comment:
due updateQuota will rebuild the quotaTree, we should add a judgement of whether we need rebuild the quotaTree actully or just update max\min or even noneed update at all in groupQuotaManager .UpdateQuota

Reply:
fixed